source/agora/node/Ledger.d

```diff
@@ -229,7 +229,10 @@ public class Ledger
     public bool onExternalized (ConsensusData data)
         @trusted
     {
-        auto block = makeNewBlock(this.last_block, data.tx_set, data.enrolls);
+        Hash random_seed = slash_man.getRandomSeed(this.getBlockHeight());
```

Comment:
One change can be made here. Take into account the `missing_validators` in the ConsensusData, and make sure their preimage is not included when calculating the random seed inside of `getRandomSeed`.

The reason why this is needed is because there can be a timing issue. A block could be externalized around the same time as another slow validator sends a preimage. So that validator could be inside of `missing_validators`, but the preimage could have been added just a second before externalization.

This is the issue I talked about before.

Reply:
> One change can be made here. Take into account the missing_validators in the ConsensusData, and make sure their preimage is not included when calculating the random seed inside of getRandomSeed.

Right, we should avoid the situation where the preimage arriving late could be used to generate the random seed. I'll fix it.